### PR TITLE
Allow connections from non-Behavior nodes

### LINKED ISSUE #57: Connections for non-pronto nodes
Maybe ct & pk?

### PR DESCRIPTION
**Pull request by phkeese**
This removes a check for a specific method on selected objects. It undoes parts of #22 and will need further review.
Closes #57 

**Progress:**
- [x] Connecting signals from non-Behavior nodes
- [x] Allowing edits of these connections
- [x] Check relevant methods where Behaviors are expected (Mostly related to Godot's Editor plugins and handling of input)

**TODO:**
- [ ] Render connections from non-Behavior nodes
- [ ] Refactor [method checks](#Duck-typing)

**Rendering of connection lines:**
Behavior nodes override their `_draw()` method to draw their own connections.
Non-Behavior nodes would need to have their connections drawn by another node.
As they are most probably going to be `CanvasItem`s, we could use the `draw()` signal (emitted after the notification, before `_draw()` is called) in combination with something like a `ConnectionDrawer` to render connections.
This will need further research.

Another idea would be to bypass `Behavior` completely for drawing connections and instead render them all every time the canvas needs to be redrawn. In my testing, rendering a `Behavior` happens only when zooming the editor and moving a `Behavior` which already has at least one connection.

**Duck-typing vs explicit type checking:**
This PR also makes changes to code relying on a passed node to have certain methods only found on `Behavior`.
This code now first checks whether the requested method actually exists.
That means, that - theoretically - any node implementing those methods would pass the test, but might not behave as expected. Maybe we should do an explicit check for `node is Behavior` instead? 
I cannot think of any potential gains by allowing [duck-typing](https://en.wikipedia.org/wiki/Duck_typing) here, since we actually expect some very specific code to be executed (Methods like `_forward_canvas_draw_over_viewport`).
Furthermore, some methods have generic names like `deselected()` which could well be part of an unrelated Class's interface.

### CODE REVIEW

File: addons/pronto/pronto.gd [hunk 1]
```diff
@@ -109,12 +109,7 @@ func _is_editing_behavior():
 	if not is_instance_valid(edited_object):
 		# edited_object might be freed after inspecting an object from a prior debugging session
 		return false
-	if not edited_object.has_method('_forward_canvas_draw_over_viewport'):
-		# edited_object is EditorDebuggerRemoteObject, which we can only use to retrieve state
-		# but not to interact with
-		# https://github.com/hpi-swa-lab/godot-pronto/pull/22
-		return false
-	return edited_object is Behavior
```
Comment (phkeese) [inline addons/pronto/pronto.gd:117]: Could this maybe be relaxed to check whether the object is a `EditorDebuggerRemoteObject`?

Comment (LinqLover) [reply addons/pronto/pronto.gd:117]: Sure, try it. But I believe that this proxy did not support that. Also, please keep the original comment :)

**Comment by tom95** [comment]
To best assess what strategy to choose for rendering connections, I was wondering what nodes you had in mind to create connections from? Do you have examples?

**Comment by phkeese** [comment]
@tom95 Mainly those inheriting from `Node2D`, but something like `AnimationPlayer` might also be interesting.

**Comment by tom95** [comment]
Do you have more concrete examples for Node2D ones? For the animationplayer to be positionable, we would likely need a wrapper node either way, probably even in the tree. Maybe this could be the way to go in general? (Just thinking out loud)

**Comment by phkeese** [comment]
@tom95 I would guess most likely other 2D-Nodes but having support for any node would be even better. This raises the question, where to draw connections from for a non-`CanvasItem` Node. I don't think the approach we took for connections **to** those node would work here, as their parent node might also have other connections and would overlap in the editor.
This might be a case for a dedicated editor panel for a Node's connection.

**Comment by tom95** [comment]
**Rendering**
* using the _draw signal sounds like a good idea. It's effectively the same as overriding _draw after all.
* ConnectionsList could be used to add a listener to get_tree().node_added on its _ready, to subscribe to _draw for the item if it's a CanvasItem.
* The remaining challenge is to store the Lines object for the Node. I imagine you could create a Dictionary of Node->Lines to do so. In get_tree().node_removed you could clear the entry.

**Duck Typing**
* I imagine a good chunk of methods will disappear from Behavior and be moved to ConnectionsList instead. Otherwise, I think I might feel more comfortable with `is Behavior` checks, for the reasons you already gave.

**Connections from Non-CanvasItems**
* Likely, the trick would just be to collect connections from non-canvasitems during render. drop_list.gd:all_children_without_position() already does a similar thing.

**Comment by tom95** [comment]
I have tested and implemented a third option now:
* For each non-Behavior node, we add a special "hidden_child" Behavior node that is only responsible for providing the Behavior API and drawing hook
* In the few places where the Behavior is directly addressed and where drawing concerns are handled, we make sure to direct all code to the hidden_child Behavior. Anywhere else, the non-Behavior node can be used directly (e.g., for connections).

It is possible that I missed some places where we need to direct the control flow to either of the two.

**Pull request closed without merging**